internal/xds/translator/authorization.go

```diff
@@ -0,0 +1,265 @@
+	// The matchers will be evaluated in order, and the first one that matches
+	// will be used to determine the action, the rest of the matchers will be
+	// skipped.
+	// If no matcher matches, the default action will be used.
+	for _, rule := range authorization.Rules {
+		// Build the IPMatcher based on the client CIDRs.
+		ipRangeMatcher := &ipmatcherv3.Ip{
+			StatPrefix: "source_ip",
+		}
+
+		for _, cidr := range rule.Principal.ClientCIDRs {
+			ipRangeMatcher.CidrRanges = append(ipRangeMatcher.CidrRanges, &configv3.CidrRange{
+				AddressPrefix: cidr.IP,
+				PrefixLen: &wrapperspb.UInt32Value{
+					Value: cidr.MaskLen,
+				},
+			})
+		}
+
+		if ipMatcher, err = anypb.New(ipRangeMatcher); err != nil {
+			return err
+		}
+
+		if sourceIPInput, err = anypb.New(&networkinput.SourceIPInput{}); err != nil {
+			return err
+		}
+
+		// Determine the action for the current rule.
+		ruleAction := allowAction
+		if rule.Action == egv1a1.AuthorizationActionDeny {
+			ruleAction = denyAction
+		}
+
+		// Add the matcher generated with the current rule to the matcher list.
+		matcherList = append(matcherList, &matcherv3.Matcher_MatcherList_FieldMatcher{
+			Predicate: &matcherv3.Matcher_MatcherList_Predicate{
+				MatchType: &matcherv3.Matcher_MatcherList_Predicate_SinglePredicate_{
+					SinglePredicate: &matcherv3.Matcher_MatcherList_Predicate_SinglePredicate{
+						Input: &cncfv3.TypedExtensionConfig{
+							Name:        "source_ip",
+							TypedConfig: sourceIPInput,
+						},
+						Matcher: &matcherv3.Matcher_MatcherList_Predicate_SinglePredicate_CustomMatch{
+							CustomMatch: &cncfv3.TypedExtensionConfig{
+								Name:        "ip_matcher",
+								TypedConfig: ipMatcher,
+							},
+						},
+					},
+				},
+			},
+			OnMatch: &matcherv3.Matcher_OnMatch{
+				OnMatch: &matcherv3.Matcher_OnMatch_Action{
+					Action: &cncfv3.TypedExtensionConfig{
+						Name:        "action",
+						TypedConfig: ruleAction,
+					},
+				},
+			},
+		})
+	}
+
+	// Set the default action.
+	defaultAction := denyAction
+	if authorization.DefaultAction == egv1a1.AuthorizationActionAllow {
+		defaultAction = allowAction
+	}
+
+	routeCfgProto := &rbacv3.RBACPerRoute{
+		Rbac: &rbacv3.RBAC{
+			Matcher: &matcherv3.Matcher{
+				MatcherType: &matcherv3.Matcher_MatcherList_{
+					MatcherList: &matcherv3.Matcher_MatcherList{
+						Matchers: matcherList,
+					},
+				},
+				OnNoMatch: &matcherv3.Matcher_OnMatch{
+					OnMatch: &matcherv3.Matcher_OnMatch_Action{
+						Action: &cncfv3.TypedExtensionConfig{
+							Name:        "default",
+							TypedConfig: defaultAction,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// If there are no matchers, the default action will be used for all requests.
+	// Setting the matcher type to nil since Proto validation will fail if the list
+	// is empty.
+	if len(matcherList) == 0 {
+		routeCfgProto.Rbac.Matcher.MatcherType = nil
+	}
+
+	if err = routeCfgProto.ValidateAll(); err != nil {
```

Comment:
id say not needed here, we are running this in https://github.com/envoyproxy/gateway/blob/2880a55b014a09deacadaa121396ea23d0a85fd4/internal/xds/types/resourceversiontable.go#L78

Reply:
We need to validate the RBACPerRoute message before converting it to an Any since the ValidateAll can't validate `Any` message.